internal/util/utils.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// todo(camilamacedo86): push this helpers to kubbuilder
+
+package util
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strings"
+)
+
+func ReplaceInFile(path, old, new string) error {
```

Comment:
As we are going to use this functions to scaffold, they should receive a `filesystem.Filesystem` and use `afero` internally instead of `ioutil` or `os`. If we use these others directly, once we implement improvements in the filesystem abstraction (for example the scaffold atomicity we talked about) these methods will fail.

Reply:
We can do that in a follow-up. This funcs or part of them exist in KB https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/utils/util.go#L63

Note that there or here we might Insert/Replace data in the files. So, IMO it can be promoted as utils from KB. 
Then, we can check the possibility to change that to use afero.

Reply:
Also see that we have a very complex code in the manifests plugin that could be simplified with that.

Reply:
But before this PR, and in the KB package you mentioned, they are used during tests after running the correspondig command call, which means that the file exists in disk.

However, here you are using these functions during scaffolding, and all scaffolding should be done through the filesystem abstraction, not directly with `os` and `ioutil` packages. Thats the point I wanted to make. I know the current machinery only allows to create fields from templates, or insert code-fragments before special markers. Do we need to improve the machinery for generalized insertion and replacement? Yes. But until that is done, SDK will have to rely on these functions, and scaffolding directly to the filesystem without using the abstraction provided should not be done.

Reply:
It is not about the file. It is about the content not be found. The same issue can happen there indeed in the tests. So, for example,  the test will pass because the required change has not been performed. 

However, I am totally fine with the improvements sugggested getting done in follow ups. the intention here was just to share the scenario.

Reply:
I'm not talking about the logic if the content is not found Camila, I'm talking about the filesystenm abstraction. We are providing one common filesystem abstraction that all `Scaffold()` **must** use. This allows us to, for example, test by saying that the underlying filesystem is an `afero.MemMapFs`. If we do this, when it gets to this scaffold, the part that uses the system abstraction will work, but the direct calls to `os` and `ioutil` will break. If you want to use these functions inside `Scaffold` and `PreScaffold` methods, you must turn those calls to `os` and `ioutil` functions to the equivalents of `afero` and pass the filesystem abstraction. If you only want to use them for tests (like until now) you don't have this problem. After moving everything to `afero`, when you call this function from tests that don't have a filesystem you just create one like so: `filesystem.Filesystem{FS: afero.NewOsFs()}`.

Reply:
Could not it be done in a follow-up? Could we not done this after we improve these implementations upstream and provide them from there? Is that a blocker for we to move forward here?

Reply:
I fear it should block, but should be pretty easy:
- Add `fs filesystem.Filesystem` as an argument.
- When calling this function from inside `Scaffold` pass the `fs` that you have there.
- When calling this function from any other place (tests) pass it a `filesystem.Filesystem(FS: afero.NewOsFs())`.
- When using `ioutil.Xxxxxx(...)` change it to `afero.Xxxxxx(fs.FS, ...)`.
- When using `os.Xxxxxx(....)` change it to `fs.FS.Xxxxxx(...)`.

Reply:
What bug/issue we will face that justify we block this PR instead of address that in follow up when your pr https://github.com/kubernetes-sigs/kubebuilder/pull/2119 get merged?